nasa/harmony-SMAP-L2-gridding-service
# DAS-2252 Initial Repository Import

## Description

This is the first import of the software to be used in the SMAP-L2-Gridding-Service.

This PR contains the software and tests to allow a user to create gridded NetCDF4-CF output data from SMAP Enhanced L2 Radiometer Half-Orbit 9 km EASE-Grid Soil Moisture data (SPL2SMP_E).

The Acceptance Criteria for this ticket are:
- [ ] Software is available in a public repository on GitHub (should be under NASA organization, but could be personal repo if the NAMS request is held up.)
- [ ] All code is properly formatted and named according to black and pep8.
- [ ] pre-commit hooks are set up and run automatically on the repository as well as for developers locally.
- [ ] Bonus points for mypy checking

To satisfy those critera:
- `pre-commit` has been enabled locally with the standard DAS `.pre-commit-config.yaml` file to enable black formatting and sort imports.
- `pre-commit.ci` has been enabled on this GitHub repository (see the checks run on [this PR](https://github.com/flamingbear/harmony-SMAP-L2-gridding-service/pull/2)).
- Mypy checking has been enabled as a first reuable action in the `run_tests_on_pull_requests.yml` action.
- You can look at the forked actions that have run https://github.com/flamingbear/harmony-SMAP-L2-gridding-service/pull/2/checks
- Code has been placed in the proper locations according to the [DAS Harmony Services General Repository Structure](https://wiki.earthdata.nasa.gov/display/SITC/DAS+Harmony+Services+General+Repository+Structure) wiki page.
- [DAS Harmony Services General Repository Structure](https://wiki.earthdata.nasa.gov/display/SITC/DAS+Harmony+Services+General+Repository+Structure) wiki page has been updated for `smap_l2_gridder`

What is not complete:
- Wrapping with harmony service built with Docker (DAS-2254)
- Building Docker Images and Testing with GitHub actions (DAS-2267)
- Connecting Service to Harmony with UMM-S and harmony PRs (DAS-2255)


## Jira Issue ID

[DAS-2252](https://bugs.earthdata.nasa.gov/browse/DAS-2252)

## Local Test Steps

- Complete a full review of the codebase.

- Download a SPL2SMP_E granule (two ways):
  + Download the .h5 file directly from Earthdata Search to be used as your input file.
  + Use Earthdata Search customization and reformat to NetCDF4-CF without any other transformations to get your `processed_granule.nc` file

- Create an isolated python 3.12 environment installing packages from the `pip_requirements.txt` and  `pip_dev_requirements.txt` file.

- run the tests:
  `❯ PYTHONPATH=. pytest .`
   *not sure why I needed to set my PYTHONPATH...*

- You can check the test coverage:
  `❯ PYTHONPATH=. pytest --cov=smap_l2_gridder --cov-report=term-missing --cov-report=html`
  *you have to install `pytest-coverage`* and `__main__.py` is not tested.

- regrid your granule. From the root directory run:
  `python -m smap_l2_gridder --input [path/to/input-file.h5] --output [path/to/output_granule.nc]`

- Validate your output against the `processed_granule.nc` from EDS. (Look on the [DAS-2252](https://bugs.earthdata.nasa.gov/browse/DAS-2252) ticket for notebook that you can use.)
  + You should be able to open both the `processed_granule.nc` and your `output_granule.nc` files with panoply.
  + Run the notebook and verify all of the data is the same with the exception of the incorrect coordinate variables in the EDS generated file. 


## PR Acceptance Checklist
* [N/A] `docker/service_version.txt` updated if publishing a release.
